### PR TITLE
Rewrite passing conventions

### PR DESCRIPTION
**Pull request by Veykril**

### CODE REVIEW

File: src/expressions.rst [hunk 1]
```diff
@@ -2101,8 +2058,8 @@ The :t:`evaluation` of a :t:`basic assignment` proceeds as follows:
    uninitialized :t:`field` of a :t:`variable`.
 
 #. :dp:`fls_hc01gtvlxba`
-   The :t:`value` of the :t:`value operand` is passed :t:`by value` into the
-   place of the :t:`assignee operand`.
+   The :t:`value` of the :t:`value operand` is :t:`passed` into the
```
Comment (kirtchev-adacore) [inline src/expressions.rst:2061]: A suggestion: change all the :t:`passed` to :t:`passed <passing convention>` since I noticed that some term references are inconsistent (passed vs [pass]ed), and it makes sense to link directly to passing conventions.

File: src/glossary.rst [hunk 2]
```diff
@@ -4321,14 +4321,23 @@ partially hygienic
 An :t:`identifier` is :dt:`partially hygienic` when it has
 :t:`mixed site hygiene`.
 
+.. _fls_wqbd5lxkddddl:
+
+passing
```
Comment (kirtchev-adacore) [inline src/glossary.rst:4326]: A suggestion: remove "passing" and stick only with "passing convention".

File: src/glossary.rst [hunk 3]
```diff
@@ -4321,14 +4321,23 @@ partially hygienic
...
 .. _fls_wqbd5lxki2al:
 
 passing convention
 ^^^^^^^^^^^^^^^^^^
 
 :dp:`fls_eqgsg8j9btic`
-A :dt:`passing convention` is a mechanism by which a :t:`value` is passed to
-and from a :t:`function`.
+A :dt:`passing convention` is the mechanism that defines how a :t:`value` is
+transferred between :t:`[place]s`.
```
Comment (kirtchev-adacore) [inline src/glossary.rst:4340]: The glossary should define what a "place" is.

File: src/ownership-and-deconstruction.rst [hunk 4]
```diff
@@ -256,71 +256,55 @@ Variable ``x`` is captured using a unique immutable borrow.
...
 
 :dp:`fls_3ztdz02efeoc`
-A :t:`value` of a :t:`by move type` is passed :dt:`by move`. Passing :t:`by
+A :t:`value` of a :t:`move type` is :t:`[pass]ed` :dt:`by move`. Passing :t:`by
 move` changes the :t:`owner` of the :t:`value`.
 
-:dp:`fls_ljc1k8ms16gp`
-Passing :dt:`by value` is either passing :t:`by copy` or passing :t:`by move`.
+:dp:`fls_konzgoybhfqm`
+A :t:`value` of a :t:`place expression` shall be :t:`[pass]ed` :t:`by move` only
```
Comment (kirtchev-adacore) [inline src/ownership-and-deconstruction.rst:289]: This sounds like a fact (tool guaranteed) rather than a rule (programmer requirement).

**Review by kirtchev-adacore — CHANGES_REQUESTED** [review]

File: src/ownership-and-deconstruction.rst (hunk 4 above)
Comment (kirtchev-adacore) [reply src/ownership-and-deconstruction.rst:289]: After discussing this live, it is in fact a rule because the programmer can control the passing convention explicitly.

File: src/expressions.rst [hunk 5]
```diff
@@ -4763,52 +4713,65 @@ A :t:`capture target` is either a :t:`variable` or a :t:`field` of a
 
 :dp:`fls_e70ywb8191h`
 The :t:`capturing environment` of a :t:`capturing expression` consists of all
-:t:`[capture target]s` that are defined outside the :t:`capturing expression`.
+:t:`captured` :t:`capture target` :t:`[value]s`.
```
Comment (kirtchev-adacore) [inline src/expressions.rst:4716]: I would rephrase this as

The :t:`capturing environment` of a :t:`capturing expression` consists of the :t:`[value]s of all :t:`captured` :t:`[capture target]s`.

File: src/glossary.rst [hunk 6]
```diff
@@ -4592,6 +4592,14 @@ excludes :t:`[range pattern]s`.
 :dp:`fls_Rj8ir4k0K811`
 See :s:`PatternWithoutRange`.
 
+.. _fls_5zjHBZMsCqJZ:
+
+place
+^^^^^
+
+:dp:`fls_uCTiUBWHMPY9`
+A :dt:`place` is a location where a :t:`value` resides at.
```
Comment (kirtchev-adacore) [inline src/glossary.rst:4601]: Just

... where a :t:`value` resides.

(without the "at")

File: src/ownership-and-deconstruction.rst [hunk 7]
```diff
@@ -258,65 +258,47 @@ Passing Conventions
...
-* :dp:`fls_olqpm32j8va6`
-  Captured by a :t:`capture expression`,
-
-* :dp:`fls_czune894326w`
-  Returned from a :t:`function`.
-
-* :dp:`fls_qi148dixkp0w`
-  **More?**
+A :t:`passing convention` is the mechanism that defines how a :t:`value` is
+transferred between :t:`[place]s`. **place is not the appropriate term here**
```
Comment (kirtchev-adacore) [inline src/ownership-and-deconstruction.rst:262]: Is "place" still not the appropriate term?

File: src/ownership-and-deconstruction.rst [hunk 8]
```diff
@@ -258,65 +258,47 @@ Passing Conventions
...
+
+:dp:`fls_konzgoybhfqm`
+A :t:`value` of a :t:`place expression` shall be :t:`passed <passing
+convention>` :t:`by move` only when it denotes:
 
-:dp:`fls_ljc1k8ms16gp`
-Passing :dt:`by value` is either passing :t:`by copy` or passing :t:`by move`.
+* :dp:`fls_4bnbv7mqod57`
+  A :t:`field` of a :t:`place expression` that can be :t:`passed <passing
+  convention>` :t:`by move` and whose type does not implement the
```
Comment (kirtchev-adacore) [inline src/ownership-and-deconstruction.rst:286]: :t:`type`

File: src/ownership-and-deconstruction.rst [hunk 9]
```diff
@@ -258,65 +258,47 @@ Passing Conventions
...
-:t:`value` is :t:`immutable`.
+* :dp:`fls_vk1xhvdaakh0`
+  A :t:`variable` which is not currently :t:`borrowed`.
 
-:dp:`fls_WW0GvbiEnyiW`
-Passing :dt:`by unique immutable reference` is passing
-:t:`by immutable reference` where it is asserted that the :t:`reference` to
-the :t:`value` is the only live :t:`reference`.
+:dp:`fls_vveEJn7lngT8`
+A :t:`value` of a :t:`value expression` is always :t:`passed <passing
```
Comment (kirtchev-adacore) [inline src/ownership-and-deconstruction.rst:296]: I prefer not to split references to terms across two lines as this makes searching hard.

File: src/patterns.rst [hunk 10]
```diff
@@ -1174,6 +1174,18 @@ deconstructor`.
 A :t:`binding` of a :t:`binding pattern` binds a matched :t:`value` to a
 :t:`name`.
 
+:dp:`fls_RViC5UEZPQUV`
+A :t:`binding` with :t:`binding mode` :dt:`by value` binds the matched :t:`value`
+by :t:[pass]ing` to the :t:`name`.
```
Comment (kirtchev-adacore) [inline src/patterns.rst:1179]: "by passing to the name" reads strangely. Did you mean "binds the matched value to a name", "binds the matched value to a name by passing the value to the place indicated by the name", "binds the matched value to the place of a name", or something similar?

**Review by kirtchev-adacore — CHANGES_REQUESTED** [review]
One hell of a cleanup!

File: src/ownership-and-deconstruction.rst (hunk 7 above)
Comment (Veykril) [reply src/ownership-and-deconstruction.rst:262]: I am still unsure about the word, given `place expression` and the like

Comment (kirtchev-adacore) [reply src/ownership-and-deconstruction.rst:262]: Fair enough.

**Review by kirtchev-adacore — APPROVED** [review]

**Comment by kirtchev-adacore** [comment]
bors merge
